File: nami/transforms/aggregator.js
/**
 * Aggregator Transform - Data aggregation and grouping
 */
class Aggregator {
    constructor(options) {
        this.groupBy = options.groupBy;
        this.aggregations = options.aggregations;
    }

    // Memory optimization has been implemented
    transform(data) {
        const groups = new Map();
        
        // Stream processing for large datasets
        for (const item of data) {
            const key = this.getGroupKey(item);
            if (!groups.has(key)) {
                groups.set(key, this.initGroup());
            }
            this.updateGroup(groups.get(key), item);
        }
        
        return Array.from(groups.values());
    }

    getGroupKey(item) {
        return this.groupBy.map(k => item[k]).join('|');
    }
}

module.exports = Aggregator;

File: nami_py/transforms/aggregator.py
"""
Aggregator transform module
"""

class Aggregator:
    """Data aggregation transform"""
    
    def __init__(self, group_by, agg_func):
        self.group_by = group_by
        self.agg_func = agg_func
    
    def process(self, data):
        """Aggregate data by groups"""
        groups = {}
        for item in data:
            key = item.get(self.group_by)
            if key not in groups:
                groups[key] = []
            groups[key].append(item)
        
        # Memory-efficient implementation
        result = []
        for key, items in groups.items():
            result.append({
                self.group_by: key,
                'value': self.agg_func(items)
            })
            # Clear processed items to prevent memory leak
            items.clear()
        
        return result
